pkg/execution/executor/executor.go

```diff
@@ -1168,6 +1232,101 @@ func (e *executor) handleGeneratorSleep(ctx context.Context, gen state.Generator
 	return err
 }
 
+func (e *executor) handleGeneratorInvokeFunction(ctx context.Context, gen state.GeneratorOpcode, item queue.Item, edge queue.PayloadEdge) error {
+	logger.From(ctx).Info().Msg("handling invoke function")
+
+	opts, err := gen.InvokeFunctionOpts()
+	if err != nil {
+		return fmt.Errorf("unable to parse invoke function opts: %w", err)
+	}
+	expires, err := opts.Expires()
+	if err != nil {
+		return fmt.Errorf("unable to parse invoke function expires: %w", err)
+	}
+
+	eventName := event.FnFinishedName
+	correlationID := item.Identifier.RunID.String() + "." + gen.ID
+	strExpr := "async.data." + consts.InvokeCorrelationId + " == '" + correlationID + "'"
+	_, err = e.newExpressionEvaluator(ctx, strExpr)
+	if err != nil {
+		return execError{err: fmt.Errorf("failed to create expression to wait for invoked function completion: %w", err)}
+	}
+
+	logger.From(ctx).Info().Interface("opts", opts).Time("expires", expires).Str("event", eventName).Str("expr", strExpr).Msg("parsed invoke function opts")
+
+	pauseID := uuid.New()
+	opcode := gen.Op.String()
+	err = e.sm.SavePause(ctx, state.Pause{
+		ID:          pauseID,
+		WorkspaceID: item.WorkspaceID,
+		Identifier:  item.Identifier,
+		GroupID:     item.GroupID,
+		Outgoing:    gen.ID,
+		Incoming:    edge.Edge.Incoming,
+		StepName:    gen.Name,
+		Opcode:      &opcode,
+		Expires:     state.Time(expires),
+		Event:       &eventName,
+		Expression:  &strExpr,
+		DataKey:     gen.ID,
+	})
+	if err != nil {
+		return err
+	}
+
+	// This should also increase the waitgroup count, as we have an
+	// edge that is outstanding.
+	//
+	// TODO: Remove with function run specific queues
+	if err := e.sm.Scheduled(ctx, item.Identifier, edge.Edge.IncomingGeneratorStep, 0, nil); err != nil {
+		return fmt.Errorf("unable to schedule function invocation: %w", err)
+	}
+
+	// Enqueue a job that will timeout the pause.
+	jobID := fmt.Sprintf("%s-%s-%s", item.Identifier.IdempotencyKey(), gen.ID, "invoke")
+	err = e.queue.Enqueue(ctx, queue.Item{
+		JobID:       &jobID,
+		WorkspaceID: item.WorkspaceID,
+		// Use the same group ID, allowing us to track the cancellation of
+		// the step correctly.
+		GroupID:    item.GroupID,
+		Kind:       queue.KindPause,
+		Identifier: item.Identifier,
+		Payload: queue.PayloadPauseTimeout{
+			PauseID:   pauseID,
+			OnTimeout: true,
+		},
+	}, expires)
+	if err == redis_state.ErrQueueItemExists {
+		return nil
+	}
+
+	// Always create an invocation event.
+	evt := event.NewInvocationEvent(event.NewInvocationEventOpts{
+		Event:         *opts.Payload,
+		FnID:          opts.FunctionID,
+		CorrelationID: &correlationID,
+	})
+
+	logger.From(ctx).Debug().Interface("evt", evt).Str("gen.ID", gen.ID).Msg("created invocation event")
+
+	if e.handleSendingEvent == nil {
```

Comment:
We should check this at the top before starting this function.  Fail early before any mutations.